trunk/ bayonne-cygwin --username wilson.tangweisheng/ccaudio2-1.0.0/utils/audiotool.1
.TH audiotool "1" "October 2004" "GNU ccAudio Tool" OST
.SH NAME
audiotool \- manipulate audio data files.
.SH SYNOPSIS
.B audiotool
\fI\-append\fR [\fI\-offset=pos\fR] \fItarget\fR \fIaudiofiles...\fR
.br
.B audiotool
\fI\-build\fR [\fI\-encoding=codec\fR] \fItarget\fR \fIaudiofiles...\fR
.br
.B audiotool
\fI\-chart\fR [\fI\-silence=value\fR] [\fI\-framing=msec\fR] \fIaudiofiles...\fR
.br
.B audiotool
\fI\-info\fR \fIaudiofile\fR
.br
.B audiotool
\fI\-notation\fR \fIaudiofile\fR [\fI``new-string''\fR]
.br
.B audiotool
\fI\-play\fR \fIaudiofiles...\fR
.br
.B audiotool
\fI-soundcard\fR [\fIindex\fR]
.br
.B audiotool
\fI\-strip\fR [\fI\-silence=value\fR] [\fI-framing=msec\fR] \fIaudiofiles...\fR
.br
.B audiotool
\fI\-trim\fR [\fI\-silence=value\fR] [\fI-framing=msec\fR] [\fI-padding=frames\fR] \fIaudiofiles...\fR
.br
.B audiotool
\fI\-version\fR
.SH DESCRIPTION
Audiotool is used to copy, manipulate, and examine the contents of audio 
files.  These audio files may be raw headerless audio data or audio stored in 
common data files such as .wav or .au files.  In the near future 
audiotool will also support packetized audio files such as .mp3 and .ogg.
.PP
.TP
\fB\-append\fR
appends the content of other audio files into an existing audio file.  
The offset to save within the existing file may be specified.  
If nessisary the content of the specified files will be re-encoded in 
the codec of the file being appended to.
.TP
\fB\-build\fR
build a brand new audio file from scratch and then copy content from 
existing audio files into it.  The encoding of the new 
audio file will either be based on the file extension, the audio file
that is being copied into it, or by specifying the encoding format 
directly.
.TP
\fB\-chart\fR
This is used to generate an energy chart profile of the audio file frame 
by frame.  A * indicates solid audio content, a ^ indicates what might 
be a lead or trailer section of partial content, and a . indicates 
relative silence.  Soon a autotrim option will be added to audiotool to 
remove silent frames.
.TP
\fB\-info\fR
The info command gives you basic information on the audio file, 
including what format it is stored as, what codec the audio data is 
encoded in, how much data is in the file, what the bit rate used is, 
etc.
.TP
\fB\-notation\fR
Notation can be used to either set or display the annotation stored in 
audio files which support annotations, such as .au and .wav files.
.TP
\fB\-play\fR
Plays the content of the audio file to the default system audio device 
(soundcard).  Play is supported on w32 and posix systems that support 
oss audio.  Future work will expand to additional platforms.
.TP
\fB\-strip\fR
Strips out all silent frames from an audio file.  Silent frames are 
those which fall below a threashold and that have no peak impulse above 
the threashold.
.TP
\fB\-trim\fR
Removes leading and trailing "silence" from a file.  The -padding option 
may be used to add additional silent frames at the end of the file after 
it has been trimed, as this may be needed for files that will later be 
used to join together phrases.
.SH BUGS
At this time audiotool does not perform and ccaudio does not 
support sample rate conversion.  Appending and building from files which 
are encoded at different sample rates could have ammusing results.
.SH AUTHOR
Written by David Sugar.
.SH "REPORTING BUGS"
Report bugs to <dyfet@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2004 Open Source Telecom Corp.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR 
PURPOSE.


trunk/ bayonne-cygwin --username wilson.tangweisheng/ccaudio2-1.0.0/utils/audiotool.cpp
// Copyright (C) 1999-2005 Open Source Telecom Corporation.
// Copyright (C) 2006-2008 David Sugar, Tycho Softworks.
//
// This file is part of GNU ccAudio2.
//
// GNU ccAudio2 is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published 
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU ccAuydio2 is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU ccAudio2.  If not, see <http://www.gnu.org/licenses/>.

#include "audiotool.h"

#ifndef	VERSION
#define	VERSION	"1.3.0"
#endif

const char *delfile = NULL;

extern "C" {

#ifdef	W32

static BOOL WINAPI down(DWORD ctrltype)
{
	if(delfile) {
		remove(delfile);
		delfile = NULL;
	}
	exit(ctrltype);
	return TRUE;
	}

#else
static void down(int signo)
{
	if(delfile)
		remove(delfile);
	exit(signo);
	}
#endif

}

static void version(void)
{
	cout << VERSION << endl;
	exit(0);
}

static void soundcard(unsigned index)
{
	AudioDevice *soundcard = Audio::getDevice(index);
	Audio::Info *info;

	if(!Audio::hasDevice(index) && !soundcard)
		cout << "Sound device inaccessible or unsupported" << endl;
	else if(!soundcard)
		cout << "Sound device unavailable" << endl;
	else {
		info = soundcard->getInfo();
		cout << "Soundcard Driver: " << info->annotation << endl;
		if(Audio::isStereo(info->encoding))
			cout << "Default Channels: 2" << endl;
		else
			cout << "Default Channels: 1" << endl;
		cout << "Default Encoding: " << Audio::getName(info->encoding) << endl;
		cout << "Default Buffers: " << info->framecount << endl;
		cout << "Default Framing: " << info->framing << "ms" << endl;
		cout << "Default Rate: " << info->rate << " samples per second" << endl;
	}
	exit(0);
}

#ifdef	CODEC_MODFLAGS
static void modflags(void)
{
	cout << CODEC_MODFLAGS << endl;
	exit(0);
}
#endif

#ifdef	AUDIO_LIBRARY
static void ldflags(void)
{
	cout << AUDIO_LIBRARY << endl;
	exit(0);
}
#endif

static void libpath(void)
{
	cout << Audio::getCodecPath() << endl;
	exit(0);
}

static void endian(void)
{
	if(__BYTE_ORDER == __BIG_ENDIAN)
		cout << "big" << endl;
	else
		cout << "little" << endl;
	exit(0);
}

#ifdef	W32
void main(int argc, char **argv)
#else
int main(int argc, char **argv)
#endif
{
	char *cp;

	if(argc < 2) {
		cerr << "use: audiotool --option [args...]" << endl;
		exit(-1);
	}
	++argv;
	cp = *argv;
	if(!strncmp(cp, "--", 2))
		++cp;

#ifdef	W32
	SetConsoleCtrlHandler(down, TRUE);
#else
	signal(SIGINT, down);
	signal(SIGTERM, down);
	signal(SIGQUIT, down);
	signal(SIGABRT, down);
#endif

	if(!stricmp(cp, "-version"))
		version();
	else if(!stricmp(cp, "-modpath") || !stricmp(cp, "-path"))
		libpath();
	else if(!stricmp(cp, "-endian"))
		endian();
	else if(!stricmp(cp, "-soundcard"))
	{
		if(*(++argv))
			soundcard(atoi(*argv));
		else
			soundcard(0);
	}
#ifdef	CODEC_MODFLAGS
	else if(!stricmp(cp, "-modflags"))
		modflags();
#endif
#ifdef	AUDIO_LIBRARY
	else if(!stricmp(cp, "-ldflags"))
		ldflags();
#endif
	else if(!stricmp(cp, "-info"))
		Tool::info(++argv);
	else if(!stricmp(cp, "-size"))
		Tool::size(++argv);
	else if(!stricmp(cp, "-notation") || !stricmp(cp, "-annotation"))
		Tool::notation(++argv);
	else if(!stricmp(cp, "-chart"))
		Tool::chart(++argv);
	else if(!stricmp(cp, "-strip"))
		Tool::strip(++argv);
	else if(!stricmp(cp, "-trim"))
		Tool::trim(++argv);
	else if(!stricmp(cp, "-build"))
		Tool::build(++argv);
	else if(!stricmp(cp, "-append"))
		Tool::append(++argv);
	else if(!stricmp(cp, "-play"))
		Tool::play(++argv);
	else if(!stricmp(cp, "-packets"))
		Tool::packetdump(++argv);

	cerr << "audiotool: " << *argv << ": unknown option" << endl;
	exit(-1);
}
